samples/order/processor.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2019 The Go Cloud Development Kit Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// A processor processes images in orders. It receives requests by
+// subscribing to the requests topic, and writes responses to the response
+// topic.
+//
+// At present, it just encodes input images into PNG format.
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"image"
+	_ "image/jpeg"
+	"image/png"
+	"log"
+	"time"
+
+	"gocloud.dev/blob"
+	_ "gocloud.dev/blob/fileblob"
+	"gocloud.dev/docstore"
+	_ "gocloud.dev/docstore/memdocstore"
+	"gocloud.dev/gcerrors"
+	"gocloud.dev/pubsub"
+	_ "gocloud.dev/pubsub/mempubsub"
+)
+
+type processor struct {
+	requestSub    *pubsub.Subscription
+	responseTopic *pubsub.Topic
+	bucket        *blob.Bucket
+	coll          *docstore.Collection
+}
+
+func (p *processor) run(ctx context.Context) error {
+	for {
+		if err := p.handleRequest(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *processor) handleRequest(ctx context.Context) error {
+	msg, err := p.requestSub.Receive(ctx)
+	if err != nil {
+		return err
+	}
+	// Ack the message because we handled it, even on error.
+	defer msg.Ack()
+
+	var req OrderRequest
+	if err := json.Unmarshal(msg.Body, &req); err != nil {
+		return err
```

Comment:
This counts as "handling" the request? I guess if it doesn't decode there's not much we can do, but worth a comment. Maybe this should return `nil` though?

Reply:
Done.